backends/platform/atari/readme.txt

```diff
@@ -0,0 +1,422 @@
+~~~~~~~~~~~~~~~~~
+
+Best of both worlds - screen tearing is avoided thanks to using of multiple
+buffers and the rendering pipeline doesn't have to wait until Vsync(). The vsync
+flag is used only to differentiate between two (very similar) modes of
+operation:
+
+1. "True triple buffering" as described in
+https://en.wikipedia.org/wiki/Multiple_buffering#Triple_buffering (vsync on)
+
+2. "Swap chain" as described in https://en.wikipedia.org/wiki/Swap_chain (vsync
+off)
+
+Pros:
+
+- best compromise between performance and visual experience
+
+- works well with both higher and lower frame rates
+
+Cons:
+
+- since three buffers are present, the buffer is always blitted into the screen
+  surface as whole, even if only one tiny little rectangle is changed (excluding
+  the cursor)
+
+- slightly irregular frame rate (depends solely on the game's complexity)
+
+- in case of extremely fast rendering in 1.), one or more buffers are
+  dropped in favor of showing only the most recent one (unlikely)
+
+- in case of extremely fast rendering in 2.), screen tearing is possible
+  because the rendering pipeline starts overwriting the buffer which is
+  currently displayed (unlikely)
+
+SuperBlitter used: yes, for rectangle blitting to screen and cursor restoration.
+Sometimes also for generic copying between buffers (see above).
+
+Triple buffering with vsync on is the default mode for this port.
+
+
+SuperVidel and SuperBlitter
+---------------------------
+
+As mentioned, this port uses SuperVidel and its SuperBlitter heavily. That
+means that if the SuperVidel is detected, it does the following:
+
+- patches all 8bpp VGA resolutions to chunky ones, rendering all C2P routines
+  useless
+
+- patches all surface addresses with OR'ing 0xA0000000, i.e. using SV RAM
+  instead of slow ST RAM (and even instead of TT RAM for allowing pure
+  SuperBlitter copying)
+
+- when SuperVidel FW version >= 9 is detected, the async FIFO buffer is used
+  instead of the slower sync blitting (where one has to wait for every
+  rectangle blit to finish). This applies only for chunky buffer -> screen
+  surfaces copy (as the generic surface copy can't rely on this behavior) but
+  despite this limitation it sometimes leads to nearly zero-cost rendering
+  and makes a *huge* difference for 640x480 fullscreen updates.
+
+
+Performance considerations/pitfalls
+-----------------------------------
+
+It's important to understand what affects performance on our limited platform
+to avoid unpleasant playing experiences.
+
+Game engines with unexpected performance hit
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A typical example from this category is the Gobliins engine (and its
+sequels). At first it looks like our machine / backend is doing something
+terribly wrong but the truth is it is the engine itself which is doing a lot of
+unnecessary redraws and updates, sometimes even before reaching the backend.
+The only real solution is to profile and fix the engine.
```

Comment:
Just FYI. This is hardly possible. The engine is mostly script-driven, and those are doing very suboptimal things.

Reply:
I don't have a precise proof but I have noticed that some engines use some kind of double buffering for example. So there is a duplicate effort: a game first renders everything in its buffer, then asks for copyRectToScreen(), I store it in another buffer and in updateScreen() I blit it to the screen.

`Screen` class is a typical example -- I would love to be able to detect when this situation occurs, something like a flag in copyRectToScreen "this is the only update for this frame, don't bother caching it".

Btw, the reports about Gobliiins being slow are quite common on Amiga forums as well.